src/workload/mod.rs

```diff
@@ -373,27 +384,44 @@ pub enum Component {
 
 #[derive(Clone)]
 pub struct Topics {
+    protocol: Protocol,
     topics: Vec<Arc<String>>,
+    partitions: usize,
     topic_dist: Distribution,
+    partition_dist: Distribution,
+    key_len: usize,
     message_len: usize,
     subscriber_poolsize: usize,
     subscriber_concurrency: usize,
+    publisher_poolsize: usize,
+    publisher_concurrency: usize,
 }
 
 impl Topics {
     pub fn new(config: &Config, topics: &config::Topics) -> Self {
         // ntopics must be >= 1
+        let protocol = config.general().protocol();
         let ntopics = std::cmp::max(1, topics.topics());
+        let partitions = topics.partitions();
         let topiclen = topics.topic_len();
         let message_len = topics.message_len();
+        let key_len = topics.key_len();
         let subscriber_poolsize = topics.subscriber_poolsize();
         let subscriber_concurrency = topics.subscriber_concurrency();
+        let publisher_poolsize = topics.publisher_poolsize();
```

Comment:
Why is publisher poolsize and concurrency relevant to workload generation?

Reply:
Because Kafka has a per workload_component publishers and subscribers. They are used here https://github.com/iopsystems/rpc-perf/pull/56/files#diff-d43550bd2ba8e3281ff197d42b301dfd7c3d574429e08fcfefedf37d7cf15358R224